pkg/clusterapi/aws/actuator.go

```diff
@@ -413,8 +464,100 @@ func (a *Actuator) Exists(cluster *clusterv1.Cluster, machine *clusterv1.Machine
 	if err != nil {
 		return false, fmt.Errorf("error getting EC2 client: %v", err)
 	}
+
+	instances, err := a.getRunningInstances(machine, client)
+	if err != nil {
+		return false, err
+	}
+	if len(instances) == 0 {
+		mLog.Debug("instance does not exist")
+		return false, nil
+	}
+
+	// If more than one result was returned, it will be handled in Update.
+	mLog.Debug("instance exists")
+	return true, nil
+}
+
+// updateStatus calculates the new machine status, checks if anything has changed, and updates if so.
+func (a *Actuator) updateStatus(machine *clusterv1.Machine, instance *ec2.Instance, mLog log.FieldLogger) error {
+
+	mLog.Debug("updating status")
+
+	// Starting with a fresh status as we assume full control of it here.
+	awsStatus := &cov1.AWSMachineProviderStatus{}
+
+	// Instance may have existed but been deleted outside our control, clear it's status if so:
+	if instance != nil {
+		awsStatus.InstanceID = instance.InstanceId
+		awsStatus.InstanceState = instance.State.Name
+		// Some of these pointers may still be nil (public IP and DNS):
+		awsStatus.PublicIP = instance.PublicIpAddress
+		awsStatus.PublicDNS = instance.PublicDnsName
+		awsStatus.PrivateIP = instance.PrivateIpAddress
+		awsStatus.PrivateDNS = instance.PrivateDnsName
+	}
+	mLog.Debug("finished calculating AWS status")
+
+	awsStatusRaw, err := controller.ClusterAPIMachineProviderStatusFromAWSMachineProviderStatus(awsStatus)
+	if err != nil {
+		mLog.Errorf("error encoding AWS provider status: %v", err)
+		return err
+	}
+
+	machineCopy := machine.DeepCopy()
+	machineCopy.Status.ProviderStatus = awsStatusRaw
+
+	if !equality.Semantic.DeepEqual(machine.Status, machineCopy.Status) {
```

Comment:
it may be more efficient to do `bytes.Equal(machine.Status.ProviderStatus.Raw, machineCopy.Status.ProviderStatus.Raw)`

Reply:
If I'm successful campaigning for IPs and DNS in the MachineStatus itself upstream, we would have to revert to this form to still catch the changes. Worth keeping as is? If not I can switch to bytes for now.

Reply:
Yup, that's fine